sysutils/pfSense-pkg-apcupsd/files/usr/local/www/widgets/widgets/apcupsd.widget.php

```diff
@@ -0,0 +1,406 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+require_once("guiconfig.inc");
+require_once("pfsense-utils.inc");
+require_once("functions.inc");
+require_once("/usr/local/www/widgets/include/apcupsd_inc.inc");
+
+if (!function_exists('compose_apc_contents')) {
+	function compose_apc_contents($widgetkey) {
+		global $user_settings;
+		
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_temp_dis_type"])) {
+			$user_settings["widgets"][$widgetkey]["apc_temp_dis_type"] = "both_deg";
+		}		
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_host_dis"])) {
+			$user_settings["widgets"][$widgetkey]["apc_host_dis"] = "no";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_temp_dis_type_var"])) {
+			$user_settings["widgets"][$widgetkey]["apc_temp_dis_type_var"] = "1";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_load_warning_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_load_warning_threshold"] = "75";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_load_critical_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_load_critical_threshold"] = "90";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_temp_warning_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_temp_warning_threshold"] = "27";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_temp_critical_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_temp_critical_threshold"] = "40";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_charge_warning_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_charge_warning_threshold"] = "50";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_charge_critical_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_charge_critical_threshold"] = "15";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_bage_warning_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_bage_warning_threshold"] = "365";
+		}
+		if (!isset($user_settings["widgets"][$widgetkey]["apc_bage_critical_threshold"])) {
+			$user_settings["widgets"][$widgetkey]["apc_bage_critical_threshold"] = "720";
+		}
+		
+		if (!(include_once "/usr/local/pkg/apcupsd.inc")) {
+			$rtnstr = "";
+			$rtnstr .= "<div class=\"panel panel-warning responsive\"><div class=\"panel-heading\"><h2 class=\"panel-title\">apcupsd not installed...</h2></div>\n";
+			$rtnstr .= "<pre>\n";
+			$rtnstr .= "Please install and configure apcupsd before using this widget.<br />\n";
+			$rtnstr .= "</pre>\n";
+			$rtnstr .= "</div>\n";
+			print($rtnstr);
+			exit(1);
+		}
+		
+		if (check_nis_running_apcupsd()) {
+			$nisip = (check_nis_ip_apcupsd() != ''? check_nis_ip_apcupsd() : "localhost");
+			$nisport = (check_nis_port_apcupsd() != '' ? check_nis_port_apcupsd() : "3551");
+
+			$ph = popen("apcaccess -h {$nisip}:{$nisport} 2>&1", "r" );
```

Comment:
1: Use full path to executable
2: All user input parameters must be run through `escapeshellarg()` before use in shell commands

Reply:
Changes submitted